### DISCUSSION TITLE
Injecting automapper in unit tests?

### DISCUSSION THREAD

**Discussion by bfahm** [nartc/mapper#416] [Q&A] [answered]
I'm trying to inject automapper when creating a testing module, but i'm getting an error because Mapper is an interface not a class.
```
const module = await Test.createTestingModule({
      providers: [
        AddressService,
        { provide: AddressesRepository, useFactory: mockAddressesRepository },
        { provide: Mapper, useFactory: mockMapper },
      ],
    }).compile();
```
I'm using the library with NestJS
How could I use the library alongside my unit tests?

**Comment by nartc** [comment, marked as answer]
@bfahm You should use `getMapperToken()` utility for the `provide` there.

If you use `InjectMapper('name_goes_here')` then use `{ provide: getMapperToken('name_goes_here'), .. }`.
If you use `InjectMapper()` then use `{ provide: getMapperToken(), ... }`

I can update docs ASAP to mention this. Thank you for asking!

**Reply by bfahm** [reply]
Thank you so much for your quick reply! Will be sure to try it out.

**Reply by bfahm** [reply]
@nartc It works, thanks again! 

Are there any approaches to use the actual Automapper (the defined profiles) in my tests without mocking them?
i.e. What is the correct approach to unit test Automapper itself in my project?

If there are any links to the docs mentioning this topic, I'd be grateful.

**Reply by micalevisk** [reply]
I guess the only way is by using the `createMapper` from `@automapper/core`

I've been using it to check if all props are being mapped properly.  

![image](https://user-images.githubusercontent.com/13461315/154809468-f6794a70-4a73-4ed5-b705-93a8be4d8a46.png)
![image](https://user-images.githubusercontent.com/13461315/154809470-fee7355e-8351-4af3-9127-bc351601e3e2.png)

**Reply by nartc** [reply]
I think that's where the `Profile` comes in. I can probably expose a static function from `AutomapperProfile` so you can get the returned value of `mapProfile()` to use in the test

```ts
{
   provide: getMapperToken(),
   useValue: createMapper({...}) // this can be a utility
}

// get the provided mapper
const testMapper = Something.get(getMapperToken());
testMapper.addProfile(SomeProfile.getProfile()); // should be the profile you want to test against
```

**Reply by micalevisk** [reply]
or even

![image](https://user-images.githubusercontent.com/13461315/154812130-6a4d698f-c6c0-4eaa-aef6-f2f514986c46.png)

**Reply by nartc** [reply]
Yeah absolutely

**Reply by bfahm** [reply]
I really appreciate the support! It worked.

**Comment by BeyondBelief96** [comment]
Hey, I've run into this issue as well and the above answer by @nartc worked for me in terms of getting the dependency resolving, but how do we go about mocking the actual map() function? I have a profile with 2 different mappings like so

`@Injectable()
export class LoanDtoMapper extends AutomapperProfile {
  constructor(@InjectMapper('loan-mapper') loanMapper: Mapper) {
    super(loanMapper);
  }

  mapProfile(): MappingProfile {
    return (mapper) => {
      mapper.createMap(LoanRequestDto, Loan);
      mapper.createMap(Loan, LoanResponseDto);
    };
  }`

and my mockLoanMapper:

export const mockLoanMapper = {
    map: jest.fn()
}


but how do I mock the map function to return the correct object given that the mapping function depends on the types specified. Like if the map function is called with the mockLoan object of type Loan, i want the mockLoanResponseDto to be returned. But if the map() function is called with the mockLoanInputDto, I want the map function to return the mockLoan object.

**Reply by nartc** [reply]
You could probably use something like `jest-when` https://www.npmjs.com/package/jest-when

**Reply by micalevisk** [reply]
another (unrelated) tip for tests in nestjs: [`@golevelup/ts-jest`](https://www.npmjs.com/package/@golevelup/ts-jest)

**Reply by BeyondBelief96** [reply]
Thanks! I'll look into those.
